src/Rules.hs

```diff
@@ -90,16 +100,21 @@ packageRules = do
     let dynamicContexts = liftM3 Context [Stage1 ..] knownPackages [dynamic]
     forM_ dynamicContexts Rules.Library.buildDynamicLib
 
-    forM_ (filter isProgram knownPackages) $
-        Rules.Program.buildProgram readPackageDb
+    Rules.Program.buildProgram readPackageDb
+
+    forM_ [Stage0 .. ] $ \stage -> do
+      -- we create a dummy context, that has the correct state, but contains
```

Comment:
I think we should define a dummy context in `Context` module. Also, I wouldn't use `base` here, but instead would create a special `Package` with an informative name (e.g. `dummy`) in case it ends up being printed somewhere to avoid confusion with actual `base`.

Reply:
Hmm, I'm not sure I see what the special package would look like, could you perhaps elaborate?

Reply:
@alpmestan Just something simple like `dummyPackage = Package "dummy"`. I think we actually already have something like this in Hadrian.

Reply:
Fixed in an upcoming commit.